frontend/src/utils.tsx

```diff
@@ -460,36 +466,72 @@ export const recommendPlaces = async (
             id: user.userId,
             geohash: user.geohash,
             friend: currentUserData.friends.includes(user.userId),
-            interestSimilarity: angleBetweenInterestVectors(
+            interestAngle: angleBetweenInterestVectors(
                 currentUserData.interests,
                 userData.interests,
             ),
         });
     }
 
+    const userLocationHistory = await getUserLocationHistory();
+
     // iterate over all nearby places, recording distance from the current user, how many other users are there, and how similar they are to the user
     for (const place of places) {
-        const userDataAtPlace = allUsersData.filter((element) =>
-            isUserAtPlace(element.geohash, place.location),
-        );
-        // get number of identical characters in place geohash and currentlocation
+        // geohash place location
         const placeGeohash = encodeBase32(
             place.location.latitude,
             place.location.longitude,
         );
 
+        // use only users at this place
+        const userDataAtPlace = allUsersData.filter((element) =>
+            areHashesClose(element.geohash, placeGeohash),
+        );
+
+        // calculate average similarity of users and number of friends
+        // if there are 0 other users there, set max difference in similarity (pi/2 radians)
+        let avgInterestAngle = Math.PI / 2;
+        let friendCount = 0;
+
+        if (userDataAtPlace.length !== 0) {
+            avgInterestAngle = 0;
+
+            for (const user of userDataAtPlace) {
+                avgInterestAngle += user.interestAngle;
+                if (user.friend) {
+                    friendCount++;
+                }
+            }
+
+            avgInterestAngle /= userDataAtPlace.length;
+        }
+
+        // calculate number of times the current user has been at this place
+        const numVisits = userLocationHistory.reduce<number>(
+            (prev, element) => {
+                if (areHashesClose(placeGeohash, element.geohash)) {
+                    return prev + 1;
+                } else {
+                    return prev;
+                }
+            },
+            0,
+        );
```

Comment:
Have you considered whether you want to take how recent the user's visits were into consideration? If the user visited a location 5 times more than a month ago do we want to treat that the same as if they visited another location 5 times in the past week?

You could play around with returning the number of visits in different recency windows such as the number of visits within the past 3 days, 7 days, 1 month, 3 months, and give more weight to the more recent ones. This would be relatively trivial since you're already storing the timestamps.

(Just something to explore - you don't have to implement this idea/approach)